### PR TITLE
Add role-based health checks to PKI Health Check

### PR DESCRIPTION
**Pull request by cipherboy**
~Based on #17750 and #17865; will be rebased once they merge.~

This adds the three role-based health checks: 

 - `role_allows_localhost`, for whether or not a role has both a non-empty `allowed_domains` list and `allow_localhost=true`.
 - `role_allows_glob_wildcards`, for whether or not a role allows both glob domains and wildcard certs, which is potentially unsafe.
 - `role_no_store_false`, for whether or not a role has roles with `no_store=false` and no auto-rebuild of CRLs, as this can cause performance issues.

### CODE REVIEW

**Comment by cipherboy** [comment]
Will be rebased once #17984 is merged.

File: command/healthcheck/pki.go
Comment (stevendpclark) [inline command/healthcheck/pki.go:247]: roleName might be a better variable name here

File: command/healthcheck/pki.go
Comment (stevendpclark) [inline command/healthcheck/pki.go:269]: Any particular reason we don't cache the fetched roles?

File: command/healthcheck/pki.go
Comment (stevendpclark) [inline command/healthcheck/pki.go:231]: nit: might be better to call out these are just role names and not the actual roles coming back?

```suggestion
func pkiFetchRoleNames(e *Executor, versionError func()) (bool, *PathFetch, []string, error) {
```

File: command/healthcheck/pki_role_allows_glob_wildcards.go
Comment (stevendpclark) [inline command/healthcheck/pki_role_allows_glob_wildcards.go:82]: We test for type checks here but no where else? Are we using this as a gate to validate we have the other keys so no need to test/validate the others?

**Review by stevendpclark — COMMENTED** [review]
Overall +1, just a few nits and a question or two to validate my understanding.

File: command/healthcheck/pki_role_allows_glob_wildcards.go
Comment (cipherboy) [reply command/healthcheck/pki_role_allows_glob_wildcards.go:82]: Ohhhhhhhhhhhhhhhhhhhhhh I'm stupid.

I was expecting `present` to mean it was in the map, not that it was safely type-cast to `bool`. D'oh!

File: command/healthcheck/pki.go
Comment (cipherboy) [reply command/healthcheck/pki.go:269]: I don't think we'd really need to do much to cache it?

The philosophy behind the cache variable is to store parsed forms.

The role JSON response (in `secret.Data`) is about exactly what we'd store here I'd think, +/- perhaps collapsing a few `[]interface{}` directly to `[]string` or similar. 

If you have ideas of data to fetch/cache here, be curious to hear it.

File: command/healthcheck/pki.go
Comment (cipherboy) [reply command/healthcheck/pki.go:231]: Should I rename all the other helpers? We have `pkiFetchIssuers` -> `pkiFetchIssuer`, `pkiFetchLeaves` -> `pkiFetchLeaf`, and here `pkiFetchRoles` -> `pkiFetchRole`. Thoughts?

File: command/healthcheck/pki_role_no_store_false.go
Comment (cipherboy) [reply command/healthcheck/pki_role_no_store_false.go]: Ah, d'oh, I copied the wrong message from the spreadsheet. Fixed.

File: command/healthcheck/pki.go
Comment (stevendpclark) [reply command/healthcheck/pki.go:231]: Your call, it would make things clearer but not a huge gain either way.

File: command/healthcheck/pki.go
Comment (stevendpclark) [reply command/healthcheck/pki.go:269]: Ah thanks for the clarification, I had missed the fact that we always internally cache the response automatically.

File: command/healthcheck/pki_role_no_store_false.go
Comment (stevendpclark) [reply command/healthcheck/pki_role_no_store_false.go]: Oops missed this myself, good catch @kitography

File: command/healthcheck/pki.go
Comment (cipherboy) [reply command/healthcheck/pki.go:231]: Yeah, the types already differ (the latter always take an issuer/serial/role reference) versus these take just two params, but I could push that in a later PR.

Comment (cipherboy) [reply command/healthcheck/pki.go:231]: Pushed in #18017

**Review by stevendpclark — APPROVED** [review]
lgtm, thanks for the tweaks.

**Comment by cipherboy** [comment]
Thanks! Merging...

**Pull request merged**
